[Custom Fields] HTML toggle

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12657 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR updates the logic of the editor to always offer a toggle to switch between the text editor and the Aztec one.

We decided to take this approach instead of showing the Aztec editor for HTML values by default for two reasons:
1. The Aztec editor can sometimes remove some parts (like the `DOCTYPE` tag) and also has some issues with inline styling, so it's best to have an explicit action from the user before using it.
2. This approach is also more flexible.

For more information, check this discussion: p1725529996823859/1724862433.885169-slack-C03L1NF1EA3

### Steps to reproduce
1. Make sure you have an order with at least one custom field.
2. Open the order on the app.
3. Tap on "View custom fields"
4. Edit or add a new field.

### Testing information
- Test the behavior of the HTML toggle.
- Confirm that the "Source" button is not shown on the Aztec toolbar.
- Type some long text on the field, and confirm scrolling works as you would expect.

### The tests that have been performed
Same as the above.

### Images/gif

https://github.com/user-attachments/assets/657ca948-701f-487e-baf6-a33e9bc3e612



- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
First pass at the toggle UI and logic

hichamboushaba committed:
Handle accessibility

hichamboushaba committed:
Move logic to the viewmodel

hichamboushaba committed:
Add support for disabling the source editor in Aztec in Compose

hichamboushaba committed:
Disable the source editor of Aztec in Custom Fields editor

hichamboushaba committed:
Add a unit test

hichamboushaba committed:
Suppress detekt issue

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Improve the animation

hichamboushaba committed:
Force a max height on the content

This avoids having an editor that spans longer than the available space making the whole UI scrollable

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/aztec/AztecEditor.kt:
```diff
@@ -280,61 +313,64 @@ private fun InternalAztecEditor(
+                }
+                aztec.sourceEditor?.doAfterTextChanged {
+                    val sourceEditor = aztec.sourceEditor
+                    if (state.isHtmlEditorEnabled || sourceEditor == null) return@doAfterTextChanged
+                    state.updateContent(sourceEditor.getPureHtml())
+                }
+
+                val focusChangeListener = OnFocusChangeListener { _, focused ->
+                    focusState.value = focused
+                }
+                aztec.visualEditor.onFocusChangeListener = focusChangeListener
+                aztec.sourceEditor?.onFocusChangeListener = focusChangeListener
+
+                viewsHolder.layout
+            },
+            update = {
+                if (aztec.visualEditor.isInCalypsoMode != calypsoMode) {
+                    aztec.visualEditor.isInCalypsoMode = calypsoMode
+                    aztec.sourceEditor?.setCalypsoMode(calypsoMode)
+                }
+
+                if (sourceEditorMinHeight != aztec.sourceEditor?.minHeight) {
+                    aztec.sourceEditor?.minHeight = sourceEditorMinHeight
+                }
+                if (minLines != -1 && minLines != aztec.visualEditor.minLines) {
+                    aztec.visualEditor.minLines = minLines
+                }
+                if (maxLines != Int.MAX_VALUE && maxLines != aztec.visualEditor.maxLines) {
+                    aztec.visualEditor.maxLines = maxLines
+                    aztec.sourceEditor?.maxLines = maxLines
+                }
+
+                if (aztec.visualEditor.label != label) {
+                    aztec.visualEditor.label = label
+                    aztec.sourceEditor?.label = label
+                }
+            },
+            modifier = modifier
+                .bringIntoViewRequester(bringIntoViewRequester)
+        )
```
All of this part wasn't touched, it's shown as edited just because it was wrapped in a `key` function.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorScreen.kt:
```diff
@@ -95,61 +136,203 @@ private fun CustomFieldsEditorScreen(
+            Column(
+                modifier = Modifier
+                    .verticalScroll(rememberScrollState())
+                    .heightIn(max = max(maxHeight, 320.dp))
```
We set the maximum height of the Column to the available `maxHeight`, unless it's less than 320.dp, this second condition is needed to avoid having an overlap between the key and value editors when `maxHeight` is small, for example when on landscape and the keyboard is shown on a phone.

JorgeMucientes reviewed (approved):
Great job @hichamboushaba. Works great. Nice toggle and really smooth transitions 🏅 

:shipit:
